content/blog/2021/08/2021-08-20-jenkins-operator-security-work-report.adoc

```diff
@@ -0,0 +1,67 @@
+---
+:layout: post
+:title: "Security Validator for Jenkins Kubernetes Operator"
+:tags:
+- gsoc21
+- jenkins
+- kubernetes
+- security
+- plugins
+authors:
+- sharmapulkit04
+description: >
+  Security validator adds an extra step of validation to detect potential security vulnerabilities in the plugins before creating a Jenkins custom resource.
+---
+
+== Background
+
+Jenkins custom resources on a Kubernetes cluster are deployed using declarative YAML configuration files; hence some of the plugins declared in these files may contain security warnings.So there is no way for the user to know other than manually checking for each on the site. 
+This project aims to add an extra step of validation before creating/updating a new Jenkins Custom Resource.
+
+== Deliverables 
+
+This project aims to add a validating admission webhook to the jenkins kubernetes operator to detect potential security vulnerabilities in the plugins before the object is created.
+
+== Dependencies 
+
+Webhooks communicate to the API server over HTTPS and use TLS. Thus, https://cert-manager.io/docs/[Jetstack/cert-manager] is used to provision TLS certificates and establish connection between Kubernetes API and webhook.  
+
+== Implementaion
+
+Operator-SDK takes care of creating a new webhook and appending it to the manager and creating handlers. Tls certificates are managed using cert-manager.
+
+* Validation Logic:
+** Proposed Implementations: Iterate through the list of plugins to be installed and fetch warnings for each plugin from the plugin center API and check if the version of that plugin has any of those warnings.  
+** Caveats: Webhooks add to API request latency,hence they should evaluate as quickly as possible thus having max allowed timeout of 30s. In the earlier approach I was fetching the security warnings from the plugin site API in the validator interface itself, and since network operations slow it was causing a timeout in the case when the number of plugins were more or internet connection was not good.
+** Updated Implementaion: Instead of fetching information for each plugin, the information of all the plugins is downloaded and cached at the start of the operator and updated periodically,thus eliminating network calls and finishing validation in less than a second.
```

Comment:
```suggestion
** Updated Implementaion: Instead of fetching information for each plugin, the information about all the plugins is downloaded and cached at the start of the operator and updated periodically, thus eliminating network calls and finishing validation in less than a second.
```